### ISSUE TITLE
Questions

### ISSUE THREAD

**Issue by XxidroxX** [d-ailin/GDN#4]
Hi, my name is Alessio and I'm writing my thesis on methods for anomaly detection in time series. I have found your paper and I would like to implement it in tensorflow but I have problems to understand the exact steps you did in this function. Can you explain it quickly?
https://github.com/d-ailin/GDN/blob/47d2663ea6558acd211ab8de56bcf660056eee29/models/GDN.py#L122

Thanks in advance for your answer.

Have a nice day!

**Comment by d-ailin** [comment]
Sure. You could ask here or contact me via email :).

**Comment by XxidroxX** [comment]
Thanks.
The first two questions are:
1) why do you reshape the input from [batch, features, time_window] to [batch*features, time_window] (line: 130) ?
2) why do you loop for "edge_index_set"? Its length is only 1 so a loop is useless I think (line 134)?

**Comment by d-ailin** [comment]
1. The reshaping here is for batch computation. You can take it as combining several batches of graphs into a larger graph so that it enables batch computation.
2. The loop here is just for some other experimental trials before. It actually can be deleted here.

**Comment by XxidroxX** [comment]
Clear, thanks.
Another question, what do the variables "gated_i" and "gated_j" mean?

**Comment by d-ailin** [comment]
gated_i, gated_j is named similarly as attr_i, attr_j in PyTorch-Geometric, they actually indicate edge_i and edge_j here.

I think you could try with this setting for SWaT
```
seed=6799
BATCH_SIZE=16
SLIDE_WIN=5
dim=64
out_layer_num=1
SLIDE_STRIDE=1
topk=15
out_layer_inter_dim=128
val_ratio=0.1
``` 

For pre-processing, you can refer to https://arxiv.org/abs/2106.06947. The first 2160 samples after downsampling would be eliminated since both water treatment systems need some time to reach stabilization when first turned on.

**Comment by flyz1** [comment]
> Clear, thanks.
> Another question, what do the variables "gated_i" and "gated_j" mean?
> 
> Im trying to reproduce the results with your code. These are the parameters for Swat dataset right?
> 
> ```
> seed=5
> BATCH_SIZE=32
> SLIDE_WIN=5
> dim=64
> out_layer_num=1
> SLIDE_STRIDE=1
> topk=15
> out_layer_inter_dim=64
> val_ratio=0.2
> decay=0
> ```
> 
> Moreover, what kind of pre-processing you did for swat dataset? Only resampling to 10s?

Hello, sorry to disturb you. I would like to ask, have you implemented the preprocessing of SWaT and WADI datasets?  Is it convenient for you to share your preprocessing code?  thank you very much!

**Comment by XxidroxX** [comment]
Hello, I have implemented only the preprocessing of SWaT. Now I dont have the code but the only operation need is the downsampling with pandas' function resample. If you need to code I can recreate in a few days.

**Comment by flyz1** [comment]
> Hello, I have implemented only the preprocessing of SWaT. Now I dont have the code but the only operation need is the downsampling with pandas' function resample. If you need to code I can recreate in a few days.

Thank you for your reply!  Since I have just started researching, I am not sure whether the code is implemented correctly by myself (in line with the requirements of the original paper). Could you please help me to recreate the code in a few days? Meanwhile, I will study more in the few days and try hard! Thank you very much!

**Comment by vgthengane** [comment]
Hi @d-ailin, I have a question regarding the training dataset. I did not understand this" Following the usual unsupervised anomaly detection formulation, the training data is assumed to consist of only normal data." in section 3.1 of a paper (https://arxiv.org/pdf/2106.06947.pdf). Is it mean we won't be adding any attack scenarios in the training dataset or something else? 

@d-ailin I am trying to reproduce the results. I have some questions;
- How do you split the training and test data? Is "WADI_14days.csv" is for training and "WADI_attackdata.csv" testing?
- Did you combine both weeks data for the WADI dataset during training? 

@XxidroxX Will, you be able to share the data preparation code here?

**Comment by XxidroxX** [comment]
Hi @vgthengane. You have to train the network with normal data only. In this way, the network is able to capture the right behaviour of the data.
Next week, when I have more free time I will publish the data_preprocessing.py

**Comment by ccskyyy** [comment]
Hi, @d-ailin May I ask what is the hyperparameter for the WADI dataset?

> gated_i, gated_j is named similarly as attr_i, attr_j in PyTorch-Geometric, they actually indicate edge_i and edge_j here.
> 
> I think you could try with this setting for SWaT
> 
> ```
> seed=6799
> BATCH_SIZE=16
> SLIDE_WIN=5
> dim=64
> out_layer_num=1
> SLIDE_STRIDE=1
> topk=15
> out_layer_inter_dim=128
> val_ratio=0.1
> ```
> 
> For pre-processing, you can refer to https://arxiv.org/abs/2106.06947. The first 2160 samples after downsampling would be eliminated since both water treatment systems need some time to reach stabilization when first turned on. We use min-max normalization as well.

**Comment by ccskyyy** [comment]
Hi @d-ailin, may I ask what is the setting for the WADI dataset? 
> gated_i, gated_j is named similarly as attr_i, attr_j in PyTorch-Geometric, they actually indicate edge_i and edge_j here.
> 
> I think you could try with this setting for SWaT
> 
> ```
> seed=6799
> BATCH_SIZE=16
> SLIDE_WIN=5
> dim=64
> out_layer_num=1
> SLIDE_STRIDE=1
> topk=15
> out_layer_inter_dim=128
> val_ratio=0.1
> ```
> 
> For pre-processing, you can refer to https://arxiv.org/abs/2106.06947. The first 2160 samples after downsampling would be eliminated since both water treatment systems need some time to reach stabilization when first turned on. We use min-max normalization as well.

**Comment by vgthengane** [comment]
I got this results for WADI dataset.
```
F1 score: 0.350208457415128
precision: 0.3788659793814433
recall: 0.3263041065482797
```
I used this setting 
```
seed=5
BATCH_SIZE=32
SLIDE_WIN=5
dim=64
out_layer_num=1
SLIDE_STRIDE=1
topk=5
out_layer_inter_dim=128
val_ratio=0.2
decay=0
```

I did min-max normalization on each column. And if all the values in the column are equal then I was getting NaN after normalization so then I replaced those with 0. Is this right way to prepare a dataset?
